worker/uniter/uniter_test.go

```diff
@@ -190,6 +190,36 @@ func (s *UniterSuite) TestUniterBootstrap(c *gc.C) {
 	})
 }
 
+func (s *UniterSuite) TestUniterRestartWithCharmDirInvalidThenRecover(c *gc.C) {
+	s.runUniterTests(c, []uniterTest{
+		ut(
+			"re-downloaded charm",
+			quickStart{},
+			createCharm{revision: 1},
+			upgradeCharm{revision: 1},
+			waitUnitAgent{
+				status: status.Idle,
+				charm:  1,
+			},
+			waitUnitAgent{
+				statusGetter: unitStatusGetter,
+				status:       status.Unknown,
+				charm:        1,
+			},
+			waitHooks{"upgrade-charm", "config-changed"},
+			verifyCharm{revision: 1},
+			verifyRunning{},
+
+			stopUniter{},
+			removeCharmDir{},
+			startUniter{}, // charm dir has been removed, restart will re-download the charm.
+			waitHooks{"start"},
```

Comment:
We should be running the upgrade-charm hook as if an upgrade had been done

Reply:
right, we need check the `upgrade-charm` for re-download charm